woocommerce/woocommerce-ios
# Order Details → Rearrange Header Elements

Closes #2279. 

Sorry for the more than 500 lines change! 🙇 300+ lines were deleted and 300+ lines were added. 
I moved a file which didn't help. 🤦 

## Design

The title and subtitle were swapped and rearranged so that:

- The subtitle is shown before the title
- The title only has the name
- The creation date is shown in the subtitle instead of the last modified date
- The order number is shown along with the creation date

Before | After 
--------|-------
![image](https://user-images.githubusercontent.com/198826/81993190-6792db00-9602-11ea-8e36-beb93616d2a3.png)        |       ![image](https://user-images.githubusercontent.com/198826/81993205-6eb9e900-9602-11ea-95b0-5319e4ace7d0.png)

After Dark | After RTL
--------|-------
![image](https://user-images.githubusercontent.com/198826/81993212-75486080-9602-11ea-8c95-ee6f2138ef69.png)|![image](https://user-images.githubusercontent.com/198826/81995389-7203a380-9607-11ea-938f-a3dc4f9b8276.png)

### Ordered Today

If the order was created on the same day, only the time will be shown.

<img src="https://user-images.githubusercontent.com/198826/81993225-7c6f6e80-9602-11ea-886a-03919b139332.png" width="320">

## Changes

This was unexpectedly trickier than I thought. 😅 

I found out that the `OrderDetailsDataSource.summaryTitle` was **tightly coupled** with the Order List cell and the Order Details header. If I changed things right there, it would affect both places! 😱 

It is also weird that we use the `OrderDetailsViewModel` in both the Order List cell and the Order Details ViewController. 🤔 See diagram below:

<img src="https://user-images.githubusercontent.com/198826/81996056-4c779980-9609-11ea-85f7-4eea7ef68262.png" width="420">


### Decoupled Order List Cell Title

I changed `OrderTableViewCell` so it uses its own function to create the title label. 

https://github.com/woocommerce/woocommerce-ios/blob/48a93a9c51fc02dab72b36c577a95182dc688955/WooCommerce/Classes/ViewRelated/Orders/Cells/OrderTableViewCell.swift#L110-L117

It currently still has a reference to `OrdersViewModel`. In the future, I think this cell should have its own `CellViewModel` that contains only the information it needs.

### Made SummaryTableViewCell More Independent

`OrderDetailsDataSource` is reaching 1000 lines. Making `SummaryTableViewCell` handle all the logic it needs to display itself makes things a bit easier to find and change IMO. The logic for what to display is encapsulated in a new [`SummaryTableViewCellViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/48a93a9c51fc02dab72b36c577a95182dc688955/WooCommerce/Classes/ViewRelated/Orders/Order%20Details/Order%20Summary%20Section/SummaryTableViewCell.swift#L10-L64) `struct`. 

The configuration in `OrderDetailsDataSource` is now a bit more simplified:

https://github.com/woocommerce/woocommerce-ios/blob/48a93a9c51fc02dab72b36c577a95182dc688955/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsDataSource.swift#L493-L504

### Unit Tests

I moved the existing `SummaryTableViewCellTests` to a folder that closely matches the `SummaryTableViewCell.swift` file. This didn't help with the PR line change. Sorry again!

I changed the test to [use mirroring](https://github.com/woocommerce/woocommerce-ios/blob/48a93a9c51fc02dab72b36c577a95182dc688955/WooCommerce/WooCommerceTests/ViewRelated/Orders/Order%20Details/Order%20Summary%20Section/SummaryTableViewCell/SummaryTableViewCellTests.swift#L99-L120) so we don't need to expose public methods in the `SummaryTableViewCell` class. 

## Testing

1. Navigate to Orders and pick an order.
2. Confirm that you see the date and the order number at the top of the billed person's name. 
3. Navigate to an order that was created today.
4. Confirm that you only see the time and the order number at the top of the billed person's name.
5. Navigate to Orders.
6. Confirm that you still see the order number and the billed person's name as the prominent (title) label. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist
